index.js

```diff
@@ -101,65 +82,148 @@ function mapChars(domainName, useSTD3, processingOption) {
+              last = i + 1;
+              continue;
+            }
+          }
+        }
+        return false;
+      }
     }
   }
 
-  return {
-    label,
-    error
-  };
+  // https://tools.ietf.org/html/rfc5893#section-2
+  if (checkBidi) {
+    let rtl;
+
+    // 1
+    if (regexes.bidiS1LTR.test(codePoints[0])) {
+      rtl = false;
+    } else if (regexes.bidiS1RTL.test(codePoints[0])) {
+      rtl = true;
+    } else {
+      return false;
+    }
+
+    if (rtl) {
+      // 2-4
+      if (!regexes.bidiS2.test(label) ||
+          !regexes.bidiS3.test(label) ||
+          (regexes.bidiS4EN.test(label) && regexes.bidiS4AN.test(label))) {
+        return false;
+      }
+    } else if (!regexes.bidiS5.test(label) ||
+               !regexes.bidiS6.test(label)) { // 5-6
+      return false;
+    }
+  }
+
+  return true;
 }
 
-function processing(domainName, useSTD3, processingOption) {
-  const result = mapChars(domainName, useSTD3, processingOption);
-  result.string = normalize(result.string);
-
-  const labels = result.string.split(".");
-  for (let i = 0; i < labels.length; ++i) {
-    try {
-      const validation = validateLabel(labels[i]);
-      labels[i] = validation.label;
-      result.error = result.error || validation.error;
-    } catch (e) {
-      result.error = true;
+function processing(domainName, options) {
+  const { checkHyphens, checkBidi, checkJoiners, processingOption } = options;
+  if (processingOption !== TRANSITIONAL && processingOption !== NONTRANSITIONAL) {
+    throw new RangeError("processingOption must be either tr46.TRANSITIONAL or tr46.NONTRANSITIONAL");
+  }
+
+  // 1. Map.
+  let { string, error } = mapChars(domainName, options);
+
+  // 2. Normalize.
+  string = string.normalize("NFC");
+
+  // 3. Break.
+  const labels = string.split(".");
+
+  // 4. Convert/Validate.
+  for (const [i, origLabel] of labels.entries()) {
+    let label = origLabel;
+    let curProcessing = processingOption;
+    if (label.startsWith("xn--")) {
+      try {
+        label = punycode.decode(label.substring(4));
+        labels[i] = label;
+      } catch (err) {
+        error = true;
+        continue;
+      }
+      curProcessing = NONTRANSITIONAL;
+    }
+
+    // No need to validate if we already know there is an error.
+    if (error) {
+      continue;
+    }
+    const validation = validateLabel(label, Object.assign({}, options, { processingOption: curProcessing }));
+    if (!validation) {
+      error = true;
     }
   }
 
   return {
     string: labels.join("."),
-    error: result.error
+    error
   };
 }
 
-module.exports.toASCII = (domainName, useSTD3, processingOption, verifyDnsLength) => {
-  const result = processing(domainName, useSTD3, processingOption);
+function toASCII(domainName, {
+  checkHyphens = false,
+  checkBidi = false,
+  checkJoiners = false,
+  useSTD3ASCIIRules = false,
+  processingOption = NONTRANSITIONAL,
+  verifyDnsLength = false
```

Comment:
Per https://w3ctag.github.io/design-principles/#casing-rules should be `verifyDNSLength`